napari/utils/colormaps/colormap.py

```diff
@@ -460,91 +640,191 @@ def _zero_preserving_modulo_numpy(
+        if values.flat[i] == to_zero:
+            out.flat[i] = 0
+        else:
+            out.flat[i] = (values.flat[i] - 1) % n + 1
+
+    return out
 
-        return out
+
+def _cast_labels_data_to_texture_dtype_direct(
+    data: np.ndarray, direct_colormap: DirectLabelColormap
+) -> np.ndarray:
+    """Convert labels data to the data type used in the texture.
+
+    In https://github.com/napari/napari/issues/6397, we noticed that using
+    float32 textures was much slower than uint8 or uint16 textures. Here we
+    convert the labels data to uint8 or uint16, based on the following rules:
+
+    - uint8 and uint16 labels data are unchanged. (No copy of the arrays.)
+    - int8 and int16 data are converted with a *view* to uint8 and uint16.
+      (This again does not involve a copy so is fast, and lossless.)
+    - higher precision integer data (u)int{32,64} are mapped to an intermediate
+      space of sequential values based on the colors they map to. As an
+      example, if the values are [1, 2**25, and 2**50], and the direct
+      colormap maps them to ['red', 'green', 'red'], then the intermediate map
+      is {1: 1, 2**25: 2, 2**50: 1}. The labels can then be converted to a
+      uint8 texture and a smaller direct colormap with only two values.
+
+    Parameters
+    ----------
+    data : np.ndarray
+        Labels data to be converted.
+    colormap : LabelColormap
+        Colormap used to display the labels data.
+
+    Returns
+    -------
+    np.ndarray
+        Converted labels data.
+    """
+    data = _convert_small_ints_to_unsigned(data)
+
+    if data.itemsize <= 2:
+        return data
+
+    return _labels_raw_to_texture_direct(data, direct_colormap)
+
+
+def _labels_raw_to_texture_direct_numpy(
+    data: np.ndarray, direct_colormap: DirectLabelColormap
+) -> np.ndarray:
+    """Convert labels data to the data type used in the texture.
+
+    This implementation uses numpy vectorized operations.
+
+    See `_cast_labels_data_to_texture_dtype_direct` for more details.
+    """
+    mapper = direct_colormap._numpy_mapper
+
+    if data.dtype.itemsize > 2:
+        data = np.clip(data, 0, mapper.shape[0] - 1)
+    return mapper[data]
+
+
+def _labels_raw_to_texture_direct_typed_dict(
+    data: np.ndarray, direct_colormap: DirectLabelColormap
+) -> np.ndarray:
+    """
+    Cast direct labels to the minimum type.
+
+    Parameters
+    ----------
+    data : np.ndarray
+        The input data array.
+    direct_colormap : DirectLabelColormap
+        The direct colormap.
+
+    Returns
+    -------
+    np.ndarray
+        The cast data array.
+    """
+    if direct_colormap.use_selection:
+        return (data == direct_colormap.selection).astype(np.uint8)
+
+    dkt = direct_colormap._get_typed_dict_mapping(data.dtype)
+    target_dtype = minimum_dtype_for_labels(
+        direct_colormap._num_unique_colors + 2
+    )
+    result_array = np.zeros_like(data, dtype=target_dtype)
```

Comment:
should this be

```suggestion
    result_array = np.full_like(data, MAPPING_OF_UNKNOWN_VALUE, dtype=target_dtype)
```
?

Reply:
(don't commit, I have a bunch of changes I want to push, just noting this down so I understand.)